config.org
# SPDX-License-Identifier: MIT AND CC-BY-SA-3.0
# SPDX-FileCopyrightText: (C) 2022-2024 nfitzen <https://github.com/nfitzen>
# SPDX-FileCopyrightText: (C) 2016-2022 Henrik Lissner.
# SPDX-FileContributor: algolix <https://stackoverflow.com/users/1402507/algolix>
#+TITLE: Doom Emacs Config
#+AUTHOR: nfitzen

Copyright \copy 2022-2024 [[https://github.com/nfitzen][nfitzen]]. \\
Copyright \copy 2016-2022 Henrik Lissner.

Most parts of this file are licensed under the
[[file:LICENSE][Expat/MIT License]]. Some are not, but I believe it's
fair use, though I note the license is
[[file:LICENSES/CC-BY-SA-3.0.txt][CC BY-SA 3.0]] anyway, with
[[https://stackoverflow.com/users/1402507/algolix][algolix]] being the
licensor for that in the [[Org Markup]] section.

Legal disclaimer out of the way, here we go.

* Table of Contents :toc:
- [[#packages][Packages]]
- [[#headers][Headers]]
- [[#clipboard][Clipboard]]
- [[#org-mode][Org Mode]]
  - [[#general][General]]
  - [[#keybindings][Keybindings]]
  - [[#org-journal][Org Journal]]
  - [[#org-habits][Org Habits]]
  - [[#org-agenda][Org Agenda]]
  - [[#org-super-agenda][Org Super Agenda]]
  - [[#org-markup][Org Markup]]
  - [[#org-latex-export][Org LaTeX Export]]
- [[#dired][Dired]]
- [[#programming][Programming]]
  - [[#web][Web]]
- [[#lsp][LSP]]
- [[#formatting][Formatting]]
  - [[#basic][Basic]]
  - [[#c][C]]
  - [[#python][Python]]
  - [[#javascript][JavaScript]]
  - [[#line-wrapping][Line Wrapping]]
- [[#themes][Themes]]
- [[#copyright][Copyright]]

* Packages
:PROPERTIES:
:header-args: emacs-lisp :tangle packages.el
:END:
These are some additional packages I install.
Will likely move these to somewhere.

#+begin_src emacs-lisp
;; -*- no-byte-compile: t -*-
#+end_src

This lets me feel like I have sensible text wrapping in documents I
can't edit as much.
#+begin_src emacs-lisp
(package! visual-fill-column)
#+end_src

For the grammar police.
#+begin_src emacs-lisp
(package! bnf-mode)
#+end_src

~org-super-agenda~ is pretty self-explanatory. My config is in the
[[Org Super Agenda]] section.
#+begin_src emacs-lisp
(package! org-super-agenda :disable t)
#+end_src

I do a bit of math stuff sometimes, so it's nice to program in R and
Sage in Emacs.
#+begin_src emacs-lisp
(package! sage-shell-mode)
(package! ob-sagemath)
(package! ess)
#+end_src

I use org-contrib as described in [[Org LaTeX Export]]. Probably
should move it there sometime.
#+begin_src emacs-lisp
(package! org-contrib)
#+end_src

I used to program Minecraft datapacks quite a lot. ~mcf-mode~
doesn't suit my needs particularly well, but the syntax highlighting
suffices well enough. I would like to hook ~lsp-mode~ into
[[https://github.com/SPYGlassMC/SPYGlass][Spyglass]] at some point.

#+begin_src emacs-lisp
(package! mcf-mode
  :recipe (:host github :repo "rasensuihei/mcf"))
#+end_src

#+begin_src emacs-lisp :tangle no
(package! minesweeper)
#+end_src
* Headers
These are some general headers. Lexical binding is the only relevant
thing right now, but I'm not even sure it's really used. Might as
well, though, since it's nicer.

#+begin_src emacs-lisp
;; -*- lexical-binding: t -*-
#+end_src
* Clipboard
I don't like Emacs constantly copying to my clipboard.
There is a solution at
https://web.archive.org/web/20150313145313/http://www.codejury.com/bypassing-the-clipboard-in-emacs-evil-mode/
But for now, I'm going the ridiculously stupid route.
...that probably isn't the Emacs way. welp.
The "+ buffer works for clipboard...

I should learn how to manage the kill ring properly, however.
It seems cool.

#+begin_src emacs-lisp
(setq! select-enable-clipboard nil)
#+end_src

* Org Mode
** General
#+begin_src emacs-lisp
(setq! org-directory "~/org/")
#+end_src
** Keybindings
#+begin_src emacs-lisp
(after! org
  (map! :map org-mode-map
        "M-q" #'fill-paragraph
        :leader
        "t p" #'+org-pretty-mode))
#+end_src
** Org Journal
#+begin_src emacs-lisp
(after! org-journal
  (setq! org-journal-enable-encryption t))
#+end_src
** Org Habits
I've recently discovered [[https://orgmode.org/manual/Tracking-your-habits.html][Org Habits]]. Pretty neat.

#+begin_src emacs-lisp
(add-to-list 'org-modules 'org-habit t)
#+end_src
** Org Agenda
#+begin_src emacs-lisp
(after! org-agenda
  (setq! org-agenda-todo-ignore-scheduled 'future
         org-agenda-tags-todo-honor-ignore-options t
         org-agenda-skip-deadline-prewarning-if-scheduled 'pre-scheduled
         org-todo-repeat-to-state t
         org-log-done 'time
         org-pretty-entities t
         ))
#+end_src
** Org Super Agenda
:PROPERTIES:
:header-args: :tangle no
:END:

#+begin_src emacs-lisp
(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq! org-super-agenda-groups '((:habit t)
                                   (:name "Events"
                                    :tag "events")
                                   (:name "Important"
                                    :priority "A")
                                   (:name "Deadlines"
                                    :tag "dl")
                                   (:name "School"
                                    :tag "school")))
  :config
  (org-super-agenda-mode))
#+end_src
** Org Markup
The following block is written by algolix in 2013, licensed under
[[https://creativecommons.org/licenses/by-sa/3.0/][CC BY-SA 3.0]].
From [[https://stackoverflow.com/a/18513349/13840781][Stackoverflow]].
Some minor changes were made from the SO post.
Tries to get org-mode to fill paragraph.
I currently don't have it working in Doom Emacs due to some
weird keybinding shenanigans.

#+begin_src emacs-lisp
(defun calc-offset-on-org-level ()
  "Calculate offset (in chars) on current level in org mode file."
  (* (or (org-current-level) 0) org-indent-indentation-per-level))

(defun my-org-fill-paragraph (&optional JUSTIFY)
  "Calculate apt fill-column value and fill paragraph."
  (let* ((fill-column (- fill-column (calc-offset-on-org-level))))
    (org-fill-paragraph JUSTIFY)))

(defun my-org-auto-fill-function ()
  "Calculate apt fill-column value and do auto-fill"
  (let* ((fill-column (- fill-column (calc-offset-on-org-level))))
    (org-auto-fill-function)))

(defun my-org-mode-hook ()
  (setq fill-paragraph-function   #'my-org-fill-paragraph
        normal-auto-fill-function #'my-org-auto-fill-function))

(add-hook! 'org-load-hook #'my-org-mode-hook)
(add-hook! 'org-mode-hook #'my-org-mode-hook)
#+end_src
** Org LaTeX Export
I write in MLA for some of my classes, so I'd like to be able to
use that. Bibliography processing is added to
~org-latex-pdf-process~ by means of repeated calls to the
processor, because LaTeX moment.

For the sake of MLA, I also want the ability to ignore headlines,
so that I can outline as I'm writing the essay. Pretty neat. This
does use the ~org-contrib~ package, however, which might be prone
to breaking. For now, it works on my machine\trade. See
[[file:packages.el][packages.el]] for the package declaration.

#+begin_src emacs-lisp
(after! ox-latex
  (setf (alist-get "mla" org-latex-classes nil nil #'string=)
        '("\\documentclass{mla}"
          ("\\section{%s}" . "\\section*{%s}")
          ("\\subsection{%s}" . "\\subsection*{%s}")
          ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
          ("\\paragraph{%s}" . "\\paragraph*{%s}")
          ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))
  (setq! org-latex-pdf-process
         '("%latex -interaction nonstopmode -output-directory %o %f"
           "%bib %b"
           "%latex -interaction nonstopmode -output-directory %o %f"
           "%bib %b"
           "%latex -interaction nonstopmode -output-directory %o %f"
           "%bib %b"
           "%latex -interaction nonstopmode -output-directory %o %f"))
  (setq! org-latex-bib-compiler "biber"))

(after! ox-extra
  (ox-extras-activate '(ignore-headlines)))
#+end_src

* Dired
#+begin_src emacs-lisp
(setq! dired-dwim-target t)
#+end_src
* Programming
** Web
#+begin_src emacs-lisp
(add-to-list 'auto-mode-alist
             '("\\.mjs" . javascript-mode))

;; (remove-hook! '(eglot-ensure) '(json-mode-hook js-mode-hook web-mode-hook))

;; (setq-hook! 'js-mode-hook js-indent-level 4)
#+end_src
* LSP
#+begin_src emacs-lisp
(add-hook! ('java-mode-hook 'python-mode-hook) 'eglot-ensure)
#+end_src
* Formatting
** Basic
I use a custom fill column function for some Apheleia formatters
because Emacs' columns are one less than the line length.

4 space indents are what I find to be the sweet spot. Tabs cause
issues with alignment, meaning a project would have to commit to
tab length. But that could be fine, I guess.

#+begin_src emacs-lisp
(add-hook! '(python-mode-hook
             js-base-mode-hook
             html-mode-hook
             css-base-mode-hook)
           :append #'apheleia-mode)

(setq! apheleia-formatters-respect-fill-column t)

(defun custom-apheleia-fill-column (fill-flag)
  "Adds 1 to fill-column for apheleia's purposes."
  (append
   (butlast (apheleia-formatters-fill-column fill-flag))
   (list (number-to-string (+ 1 fill-column)))))
#+end_src
** C
I prefer something approximating the
[[https://docs.kernel.org/process/coding-style.html][Linux coding style]],
but with 4 spaces instead of tabs. This generally applies
everywhere.

As a hack, ~func-decl-cont~ is set to 0 in C because it's
meaningless and will allow braces after function declarations to
be nicer.

I tried to use Emacs indent for C in Apheleia, but it's not
playing nice, so I'm rebinding the ~format-buffer~ key as a hack.

#+begin_src emacs-lisp
(c-add-style "my-misc-style" '("linux"
                               (c-basic-offset . 4)
                               (indent-tabs-mode . nil)))
(c-add-style "my-c-style" '("my-misc-style"
                            (c-offsets-alist
                             (func-decl-cont . 0))))
(setf (alist-get 'c-mode c-default-style) "my-c-style"
      (alist-get 'other c-default-style) "my-misc-style")

;; NOTE: adapted from apheleia-indent-lisp-buffer by me.
;; Copyright (C) 2019-2022 Radian LLC and Contributors.
;; Also under Expat/MIT License.
;; FIXME: doesn't format, instead prints a wrong type error.
(cl-defun a-reindent-buffer
    (&key buffer scratch callback &allow-other-keys)
  "Formatter for C indent with Apheleia."
  ;; (print "test")
  (with-current-buffer scratch
    (setq-local indent-line-function
                (buffer-local-value 'indent-line-function buffer))
    (funcall (with-current-buffer buffer major-mode))
    (when c-style-variables-are-local-p
      (dolist (var c-style-variables)
        (make-local-variable var)
        (set var (buffer-local-value var buffer))))
    (indent-region (point-min) (point-max))
    (funcall callback)))

;; (after! apheleia
;;   (setf (alist-get 'c-mode apheleia-mode-alist) 'c-indent
;;         (alist-get 'cc-mode apheleia-mode-alist) 'c-indent
;;         (alist-get 'c-indent apheleia-formatters) #'a-reindent-buffer))

;; copied from https://www.emacswiki.org/emacs/ReformatBuffer.
;; I believe that the merger doctrine applies because this function
;; is so simple.
(defun reindent-buffer ()
  "Reindents current buffer."
  (interactive)
  (save-excursion
    (indent-region (point-min) (point-max))))

(map! :map c-mode-map
      :nv "SPC c F" #'reindent-buffer)
#+end_src
** Python
#+begin_src emacs-lisp
(after! apheleia
  (setf (alist-get 'black apheleia-formatters)
        '(
          "black"
          (when
              (apheleia-formatters-extension-p "pyi")
            "--pyi")
          (custom-apheleia-fill-column "--line-length")
          "-")))
#+end_src
** JavaScript
Disable Apheleia

#+begin_src emacs-lisp
(setq! js-indent-level 4
       web-mode-indent-style 4
       js2-basic-offset 4)

(setq-default js-indent-level 4
              web-mode-indent-style 4
              js2-basic-offset 4)

;; (add-hook 'rjsx-mode-hook (lambda () (apheleia-mode nil)))
#+end_src
** Line Wrapping
#+begin_src emacs-lisp
(setq-default fill-column 78
              display-fill-column-indicator-column 80)

(setq-hook! 'text-mode-hook
  fill-column 70
  display-fill-column-indicator-column 72)

(add-hook! 'prog-mode-hook :append #'display-fill-column-indicator-mode)
(add-hook! 'markdown-mode-hook :append
           #'display-fill-column-indicator-mode)

(defun enable-multiline-block ()
  (if comment-multi-line
      (set-variable 'comment-style 'extra-line)))
(add-hook! 'prog-mode-hook :append 'enable-multiline-block)
#+end_src
* Themes
Miscellaneous theming and style.

#+begin_src emacs-lisp
(setq! doom-theme 'doom-one)
(setq! display-line-numbers-type 'relative)

(map! :leader
      "t c" #'display-fill-column-indicator-mode)
#+end_src
* Copyright
I am a bit of a stickler for copyright. I want to be able to insert
copyright headers into my code. The following is how I tend to do
it, which you can see at the top of this file. Minor alterations
sometimes need to be made, since I drop the "(C)" when licensing
under CC0-1.0 to have only the minimal elements of the REUSE
guidelines while trying to minimize the chance it gets seen as a
copyright notice (whereas when I want a copyright, I want to
maximize the chance under the guidelines).

#+begin_src emacs-lisp
(setq!
 custom-copyright-name "nfitzen"
 custom-copyright-contact "https://github.com/nfitzen")

(define-skeleton custom-copyright-insert
  "Copyright with SPDX header"
  "License ID: "
  comment-start
  "SPDX-License-Identifier: "
  str
  comment-end
  \n
  comment-start
  "SPDX-FileCopyrightText: (C) "
  `(format-time-string "%Y")
  " "
  custom-copyright-name
  " <"
  custom-copyright-contact
  ">"
  comment-end
  \n)

(map! :leader
      :desc "Insert copyright" "l i" #'custom-copyright-insert)
#+end_src


config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
;; SPDX-License-Identifier: MIT
;; SPDX-FileCopyrightText: (C) 2022-2023 nfitzen <https://github.com/nfitzen>
;; Copyright (c) 2016-2022 Henrik Lissner.

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

;; I don't like Evil mode constantly copying to my clipboard.
;; There is a solution at
;; https://web.archive.org/web/20150313145313/http://www.codejury.com/bypassing-the-clipboard-in-emacs-evil-mode/
;; But for now, I'm going the ridiculously stupid route.
;; ...that probably isn't the Emacs way. welp.
;; The "+ buffer works for clipboard...

(setq! select-enable-clipboard nil)

;; org mode

;; org export
(setq org-export-date-timestamp-format "%x")

(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-super-agenda-groups '(
                                  (:name "Today"
                                   :time-grid t
                                   :scheduled today)
                                  (:name "Important"
                                   :priority "A")
                                  (:name "Deadlines"
                                   :tag "dl")
                                  (:name "School"
                                   :tag "school")
                                  ))
  :config
  (org-super-agenda-mode))

(after! org-journal
  (setq org-journal-enable-encryption t))

(after! org-agenda
  (setq org-agenda-todo-ignore-scheduled 'future
        org-agenda-tags-todo-honor-ignore-options t
        org-agenda-skip-deadline-prewarning-if-scheduled 'pre-scheduled
        org-todo-repeat-to-state t
        org-log-done 'time
        org-pretty-entities t
        ))

;; (add-hook! 'org-agenda-mode-hook :append #'org-super-agenda-mode)

;; org markup

;; the following block is (C) 2013 algolix, licensed under CC BY-SA 3.0.
;; https://stackoverflow.com/a/18513349/13840781
;; BEGIN BLOCK
(defun calc-offset-on-org-level ()
  "Calculate offset (in chars) on current level in org mode file."
  (* (or (org-current-level) 0) org-indent-indentation-per-level))

(defun my-org-fill-paragraph (&optional JUSTIFY)
  "Calculate apt fill-column value and fill paragraph."
  (let* ((fill-column (- fill-column (calc-offset-on-org-level))))
    (org-fill-paragraph JUSTIFY)))

(defun my-org-auto-fill-function ()
  "Calculate apt fill-column value and do auto-fill"
  (let* ((fill-column (- fill-column (calc-offset-on-org-level))))
    (org-auto-fill-function)))

(defun my-org-mode-hook ()
  (setq fill-paragraph-function   #'my-org-fill-paragraph
        normal-auto-fill-function #'my-org-auto-fill-function))

(add-hook! 'org-load-hook #'my-org-mode-hook)
(add-hook! 'org-mode-hook #'my-org-mode-hook)
;; END BLOCK

(setq! dired-dwim-target t)

;; LSP and formatting
(add-hook! '(python-mode-hook
             js-base-mode-hook
             html-mode-hook
             css-base-mode-hook)
           :append #'apheleia-mode)

(setq! apheleia-formatters-respect-fill-column t)

(setq-default indent-tabs-mode nil
              c-basic-offset 4
              tab-width 8)

;; python
(setq! lsp-pyright-python-executable-cmd "python3")
(after! poetry
  (remove-hook 'python-mode-hook #'poetry-tracking-mode)
  (poetry-tracking-mode 1))

;; Line wrapping

(setq-default fill-column 78
              display-fill-column-indicator-column 80)

(add-hook! 'text-mode-hook :append :local
  (setq fill-column 70
        display-fill-column-indicator 72))

(add-hook! 'prog-mode-hook :append #'display-fill-column-indicator-mode)
(add-hook! 'markdown-mode-hook :append #'display-fill-column-indicator-mode)

(defun enable-multiline-block ()
  (if comment-multi-line
      (set-variable 'comment-style 'extra-line)))
(add-hook! 'prog-mode-hook :append 'enable-multiline-block)

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-one)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")


;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(map! :leader
      :desc "Show LSP docs" "c h" #'lsp-describe-thing-at-point
      "t c" #'display-fill-column-indicator-mode)

(map! :map org-mode-map
      :leader
      "t p" #'+org-pretty-mode)

(map! :after org :map org-mode-map "M-q" #'fill-paragraph)
